src/core/dense_set.cc

```diff
@@ -0,0 +1,424 @@
+      } else {
+        DensePtr node = Unlink(curr);
+        PushFront(entries_.begin() + bid, node);
+        entries_[bid].ClearDisplaced();
+      }
+    }
+  }
+}
+
+bool DenseSet::AddInternal(void* ptr) {
+  uint64_t hc = Hash(ptr);
+
+  if (entries_.empty()) {
+    capacity_log_ = kMinSizeShift;
+    entries_.resize(kMinSize);
+    uint32_t bucket_id = BucketId(hc);
+    auto e = entries_.begin() + bucket_id;
+    obj_malloc_used_ += PushFront(e, ptr);
+    ++size_;
+    ++num_used_buckets_;
+
+    return true;
+  }
+
+  // if the value is already in the set exit early
+  uint32_t bucket_id = BucketId(hc);
+  if (Find(ptr, bucket_id) != nullptr) {
+    return false;
+  }
+
+  DCHECK_LT(bucket_id, entries_.size());
+
+  // Try insert into flat surface first. Also handle the grow case
+  // if utilization is too high.
+  for (unsigned j = 0; j < 2; ++j) {
+    ChainVectorIterator list = FindEmptyAround(bucket_id);
+    if (list != entries_.end()) {
+      obj_malloc_used_ += PushFront(list, ptr);
+      if (std::distance(entries_.begin(), list) != bucket_id) {
+        list->SetDisplaced(std::distance(entries_.begin() + bucket_id, list));
+      }
+      ++num_used_buckets_;
+      ++size_;
+      return true;
+    }
+
+    if (size_ < entries_.size()) {
+      break;
+    }
+
+    Grow();
+    bucket_id = BucketId(hc);
+  }
+
+  DCHECK(!entries_[bucket_id].IsEmpty());
+
+  /**
+   * Since the current entry is not empty, it is either a valid chain
+   * or there is a displaced node here. In the latter case it is best to
+   * move the displaced node to its correct bucket. However there could be
+   * a displaced node there and so forth. Keep a stack of nodes that are displaced
+   * until we find a displaced node which hashes to an empty bucket.
+   * Then unwind the stack of pending displacements until each node is in
+   * its correct bucket
+   */
+
+  DensePtr to_insert(ptr);
+  while (!entries_[bucket_id].IsEmpty() && entries_[bucket_id].IsDisplaced()) {
+    DensePtr unlinked = PopPtrFront(entries_.begin() + bucket_id);
+
+    PushFront(entries_.begin() + bucket_id, to_insert);
+    to_insert = unlinked;
+    bucket_id -= unlinked.GetDisplacedDirection();
+  }
+
+  if (!entries_[bucket_id].IsEmpty()) {
+    ++num_chain_entries_;
+  }
+
+  ChainVectorIterator list = entries_.begin() + bucket_id;
+  PushFront(list, to_insert);
+  obj_malloc_used_ += ObjectAllocSize(ptr);
+  DCHECK(!entries_[bucket_id].IsDisplaced());
+
+  ++size_;
+  return true;
+}
+
+auto DenseSet::Find(const void* ptr, uint32_t bid) const -> const DensePtr* {
+  // first look for displaced nodes since this is quicker than iterating a poential long chain
+  if (bid && Equal(entries_[bid - 1], ptr)) {
+    return &entries_[bid - 1];
+  }
+
+  if (bid + 1 < entries_.size() && Equal(entries_[bid + 1], ptr)) {
+    return &entries_[bid + 1];
+  }
+
+  // if the node is not displaced, search the correct chain
+  const DensePtr* curr = &entries_[bid];
```

Comment:
i know it works for you code-wise. I wonder if that's indeed a good code order for searching. 

my intuition is that if the preference is to insert into `bid` first, then if you search **first** in `bid` flat entry, your expected hit-rate at the first branch will be highest.